### ISSUE TITLE
Support for more variants

### ISSUE THREAD

**Issue by jcw** [platformio/builder-framework-libopencm3#1]
Is libopencm3 still supported on PlatformIO? There are a lot STM32 µCs which are not listed.
From the README on GitHub:

> ST STM32F0xx/F1xx/F2xx/F30x/F37x/F4xx/F7xx/L0xx/L1xx/L4xx series

E.g. the F429, F7xx, L0xx, and L4xx. See also the [examples](https://github.com/libopencm3/libopencm3-examples/tree/master/examples/stm32).

Apps built with libopencm3 have little runtime overhead (an empty main is < 1K), yet with a lot support for different devices, including USART and USB. The well-known Black Magic Probe is based on libopencm3.

Given that this repository only contains a single file, how does one include more variants? Is it simply a matter or re-running that script and re-registering its results with PIO?

**Comment by ivankravets** [comment]
Could you provide PR with a support for other MCUs?

**Comment by jcw** [comment]
I’ll see what I can do. Still novice in PIO-land, especially on the fw-/lib-developer side.

**Comment by jcw** [comment]
Baby steps: I see `~/.platformio/packages/framework-libopencm3/lib/stm32/l0/` in my install of PIO.
I can make changes there, which would let me build for more boards.
But where do I make changes to be able to create a PR? Where is the actual repository for PIO?

UPDATE: I'm starting to understand the way `~/.platformio/` is set up. Some Q's remain:
* do I just replace `packages/framework/libopencm3` with a recent github clone or my fork of it?
* same Q for `platforms/ststm32` - do I just move it out of the way and put my version there?

UPDATE #2: progress - I copied from .platformio/platforms/ststm32/boards/ to my project boards/, then added libopencm3 to the "frameworks" hash (tried with both L011K4 and L053R8), but now I get this error:

```
$ pio run -v
[Sat Mar 17 23:36:28 2018] Processing try (platform: ststm32; board: nucleo_l053r8; framework: libopencm3)
------------------------------------------------------------------------------------------
PLATFORM: ST STM32 > ST Nucleo L053R8
SYSTEM: STM32L053R8T6 48MHz 8KB RAM (64KB Flash)
DEBUG: CURRENT(stlink) ON-BOARD(stlink) EXTERNAL(blackmagic, jlink)
KeyError: "Invalid board option 'build.core'":
File "/Users/jcw/.platformio/penv/lib/python2.7/site-packages/platformio/builder/main.py", line 162:
env.SConscript("$BUILD_SCRIPT")
File "/Users/jcw/.platformio/packages/tool-scons/script/../engine/SCons/Script/SConscript.py", line 541:
return _SConscript(self.fs, *files, **subst_kw)
File "/Users/jcw/.platformio/packages/tool-scons/script/../engine/SCons/Script/SConscript.py", line 250:
exec _file_ in call_stack[-1].globals
File "/Users/jcw/.platformio/platforms/ststm32/builder/main.py", line 127:
target_elf = env.BuildProgram()
File "/Users/jcw/.platformio/packages/tool-scons/script/../engine/SCons/Environment.py", line 224:
return self.method(*nargs, **kwargs)
File "/Users/jcw/.platformio/penv/lib/python2.7/site-packages/platformio/builder/tools/platformio.py", line 72:
env.BuildFrameworks(env.get("PIOFRAMEWORK"))
File "/Users/jcw/.platformio/packages/tool-scons/script/../engine/SCons/Environment.py", line 224:
return self.method(*nargs, **kwargs)
File "/Users/jcw/.platformio/penv/lib/python2.7/site-packages/platformio/builder/tools/platformio.py", line 276:
SConscript(env.GetFrameworkScript(f))
File "/Users/jcw/.platformio/packages/tool-scons/script/../engine/SCons/Script/SConscript.py", line 604:
return method(*args, **kw)
File "/Users/jcw/.platformio/packages/tool-scons/script/../engine/SCons/Script/SConscript.py", line 541:
return _SConscript(self.fs, *files, **subst_kw)
File "/Users/jcw/.platformio/packages/tool-scons/script/../engine/SCons/Script/SConscript.py", line 250:
exec _file_ in call_stack[-1].globals
File "/Users/jcw/.platformio/platforms/ststm32/builder/frameworks/libopencm3/libopencm3.py", line 145:
if env.BoardConfig().get("build.core") == "tivac":
File "/Users/jcw/.platformio/penv/lib/python2.7/site-packages/platformio/managers/platform.py", line 658:
raise KeyError("Invalid board option '%s'" % path)
=============================== [ERROR] Took 0.38 seconds ===============================
```

I can see the ref to `boards.core` in `libopencm3.py`, but don't understand why it bails out.

Just a wild shot ... has the libopencm3.py script been updated for PIO 3?

**Comment by jcw** [comment]
How is the content of `~/.platformio/packages/framework-libopencm3/` determined?
It's lagging behind the current version at https://github.com/libopencm3/libopencm3

**Comment by ivankravets** [comment]
See https://bintray.com/platformio/dl-packages/framework-libopencm3/1.1.0

**Comment by ivankravets** [comment]
More details https://github.com/libopencm3/libopencm3/issues/896#issuecomment-386325469

**Comment by mkarklinsScout** [comment]
Has there been any progress on this issue?

**Comment by JackDesBwa** [comment]
I just start to try PlatformIO and also would like to use it with libopencm3 for a STM32 chip.
The instructions given in this thread are pretty clear so that I easily managed to use the "master" version of libopencm3 with my MCU through PlatformIO, but it is quite a burden and I would like to help to make it mainstream.

However, I am not familiar at all with scons, thus my implementation is very bad for the moment. The basic idea is this one:
```(python)
def ldscript_genlink():
    mcu = env.BoardConfig().get("build.mcu")
    defs = exec_command([
        "python",
        join(FRAMEWORK_DIR, 'scripts', 'genlink.py'),
        join(FRAMEWORK_DIR, 'ld', 'devices.data'),
        mcu,
        'DEFS'
        ], cwd=FRAMEWORK_DIR)['out']
    buildldcmd = ["arm-none-eabi-gcc"]
    buildldcmd.extend(defs.split(' '))
    buildldcmd.append('-P')
    buildldcmd.append('-E')
    buildldcmd.append(join(FRAMEWORK_DIR, 'ld', 'linker.ld.S'))
    ldscript = exec_command(buildldcmd, cwd=FRAMEWORK_DIR)['out']
    ldpath = "." + mcu + ".ld"
    with open(ldpath, 'w') as f:
        f.write(ldscript)
    return ldpath
```
This piece of code is meant to replace current `find_ldscript` or complement it with a test to `join(FRAMEWORK_DIR, 'scripts', 'genlink.py')` file presence. It generates the ld script and writes it to a file.

Problems:
- arm-none-eabi-gcc is the one on my system and not from the PlatformIO toolchain
- ldpath is not in the build directory
- current libopencm3 downloaded by PlatformIO is too old (I downloaded with git and added a package.json in it for my tests)

This is a baby step (which works on my system) but it might help to implement it more properly.
Is there someone who can help to solve the listed problems I saw?
